### ISSUE TITLE
Unexpected token error at code with optional chaining when using Typescript

### ISSUE THREAD

**Issue by inoyakaigor** [webpack/webpack#10227]
<s>UPD for everyone who will find this issue. For fix this error acornjs/acorn#891 and acornjs/acorn#890 must be merged.
UPD2 we all awaiting while estree/estree/pull/204 will be merged</s>

UPD3 fix for optional chaining landed into acorn 7.3.0 but webpack@4 uses acorn 6.x. so fix of this issue will be in webpack@5: https://github.com/webpack/webpack/issues/10227#issuecomment-642734920 

### For Babel
Add `@babel/plugin-proposal-optional-chaining` and `@babel/plugin-proposal-nullish-coalescing-operator` plugins in your config

### For Typescript
Do not upgrade to version 3.8 or set compiler option `target` to `es2019`

### Yarn/NPM level
See the https://github.com/webpack/webpack/issues/10227#issuecomment-653408484

---

**Do you want to request a *feature* or report a *bug*?**
I don't know is it a bug or feature

**What is the current behavior?**

Get error "Unecpected token"

**What is the expected behavior?**

Build bundle successfull

**Description**
When I set in `tsconfig.json` `target: "ESNext"` TS keeps my code with optional chaining (and nullish coalescing)  as is. But I've got an error
```
Module parse failed: Unexpected token (2:10)
File was processed with these loaders:
 * ./node_modules/ts-loader/index.js
You may need an additional loader to handle the result of these loaders.
| var a = { b: 2 };
> var c = a?.b;
```

I made a minimum reproduce repo [inoyakaigor/webpack-typescript-optional-chaining](/inoyakaigor/webpack-typescript-optional-chaining)

How I can avoid this error?

**Comment by evilebottnawi** [comment]
It looks like you just deleted our lovely crafted [issue template](https://raw.githubusercontent.com/webpack/webpack/master/.github/ISSUE_TEMPLATE.md). It was there for good reasons. Please help us solving your issue by answering the questions asked in this template. I'm closing this. Please either update the issue with the template and reopen, or open a new issue.

**Comment by inoyakaigor** [comment]
@evilebottnawi Closing the issue because I removed the template looks like excessive bureaucracy. 

However, I updated the description.

**Comment by mc0** [comment]
@evilebottnawi looks like @inoyakaigor updated the description and also provided a reproduction repo. I think we're hitting this same issue with optional chaining.

**Comment by ianschmitz** [comment]
For others that stumble on this issue, it looks like we're waiting on https://github.com/acornjs/acorn/pull/891 and https://github.com/acornjs/acorn/pull/890 before we stop transpiling this syntax via babel/tsc.

**Comment by ospfranco** [comment]
it looks like the mentioned issues on acorn have been merged, any news?

**Comment by yangmingshan** [comment]
@evilebottnawi Can we reopen this issue? Since this is a unresolved problem, it will help people find answers and let people track the progress.

**Comment by inoyakaigor** [comment]
Good news everyone! Acorn 7.3.0 with optional chaining support just released 2 hours ago!

**Comment by ianschmitz** [comment]
Unfortunately webpack 4 uses acorn 6.x - not sure if we'll get this before webpack 5. Fingers crossed!

**Comment by evilebottnawi** [comment]
Yes, it will be part of webpack@5

**Comment by inoyakaigor** [comment]
I just tried to make a junction link _(symbolic in terms of *nix OSes)_ ./node_modules/webpack/node_modules/acorn a junction link to ./node_modules/acorn and everything build ok.

**Comment by OnurGvnc** [comment]
as a temporary solution;
https://github.com/OnurGvnc/acorn-with-stage3

`package.json`:

```json
{
  ...
  "resolutions": {
    "acorn": "npm:acorn-with-stage3"
  },
  "dependencies": {

```

**Comment by sboisse** [comment]
> as a temporary solution;
> https://github.com/OnurGvnc/acorn-with-stage3
> 
> `package.json`:
> 
> ```json
> {
>   ...
>   "resolutions": {
>     "acorn": "npm:acorn-with-stage3"
>   },
>   "dependencies": {
> ```

Maybe I am missing something, but doing this did not fix the issue for me. I still get the same unexpected token at each use of optional chaining I have in my code.

**Comment by fsevenm** [comment]
Based on @sboisse temporary solution, in Yarn we just need resolution part,
```json
"resolutions": {
    "acorn": "8.0.1"
},
```
After that, just run `yarn install`.

**Comment by evilebottnawi** [comment]
webpack uses `acorn@8`

**Comment by fsevenm** [comment]
Just the case, if we're using a library/loader that doesn't support webpack@5 yet.

**Comment by slavafomin** [comment]
It occurs that optional chaining won't be supported in Webpack 4, so the best course of action would be to update to Webpack 5 or use Babel for transpilation.

**Comment by Keavon** [comment]
What's the status on this? Why is it closed?

**Comment by sshubhamk** [comment]
I'm getting the same error-

```
Module parse failed: Unexpected token (363:46)
File was processed with these loaders:
  ./node_modules/ts-loader/index.js
You may need an additional loader to handle the result of these loaders.
```


tsconfig.json-
![image](https://user-images.githubusercontent.com/52405668/118275944-b0ff5a00-b4e4-11eb-9e00-4b6cc7a9d34c.png)

**Comment by jchnxu** [comment]
as a temporary solution;
https://github.com/OnurGvnc/acorn-with-stage3

`package.json`:

```json
{
  ...
  "resolutions": {
    "acorn": "npm:acorn-with-stage3"
  },
  "dependencies": {

```

_Originally posted by @OnurGvnc in https://github.com/webpack/webpack/issues/10227#issuecomment-653408484_

Thanks it helped solve my problem. However my problem is weird because optional chaining is already supported in my own files, the only place it breaks is in a nested dependency I just added node_modules/xxxx/node_modules/yyyy. I'm still not sure why it happened.

**Comment by Patrick-particle16** [comment]
https://reactdev.space/questions/the-error-message-encountered-while-using-webpack-with-typescript-and-react-is-unexpected-token-reactcomponent



Sift through this explanation
